wpimath/src/main/java/edu/wpi/first/math/geometry/Pose3d.java

```diff
@@ -258,51 +207,8 @@ public Pose3d exp(Twist3d twist) {
    * @return The twist that maps this to end.
    */
   public Twist3d log(Pose3d end) {
-    // Implementation from Section 3.2 of https://ethaneade.org/lie.pdf
-    final var transform = end.relativeTo(this);
-
-    final var rvec = transform.getRotation().getQuaternion().toRotationVector();
-
-    final var omega = rotationVectorToMatrix(rvec);
-    final var theta = rvec.norm();
-    final var thetaSq = theta * theta;
-
-    double C;
-    if (Math.abs(theta) < 1E-7) {
-      // Taylor Expansions around θ = 0
-      // A = 1/1! - θ²/3! + θ⁴/5!
-      // B = 1/2! - θ²/4! + θ⁴/6!
-      // C = 1/6 * (1/2 + θ²/5! + θ⁴/7!)
-      // sources:
-      // A:
-      // https://www.wolframalpha.com/input?i2d=true&i=series+expansion+of+Divide%5Bsin%5C%2840%29x%5C%2841%29%2Cx%5D+at+x%3D0
-      // B:
-      // https://www.wolframalpha.com/input?i2d=true&i=series+expansion+of+Divide%5B1-cos%5C%2840%29x%5C%2841%29%2CPower%5Bx%2C2%5D%5D+at+x%3D0
-      // C:
-      // https://www.wolframalpha.com/input?i2d=true&i=series+expansion+of+Divide%5B1-Divide%5BDivide%5Bsin%5C%2840%29x%5C%2841%29%2Cx%5D%2C2Divide%5B1-cos%5C%2840%29x%5C%2841%29%2CPower%5Bx%2C2%5D%5D%5D%2CPower%5Bx%2C2%5D%5D+at+x%3D0
-      C = 1 / 12.0 + thetaSq / 720 + thetaSq * thetaSq / 30240;
-    } else {
-      // A = sin(θ)/θ
-      // B = (1 - cos(θ)) / θ²
-      // C = (1 - A/(2*B)) / θ²
-      double A = Math.sin(theta) / theta;
-      double B = (1 - Math.cos(theta)) / thetaSq;
-      C = (1 - A / (2 * B)) / thetaSq;
-    }
-
-    final var V_inv =
-        Matrix.eye(Nat.N3()).minus(omega.times(0.5)).plus(omega.times(omega).times(C));
-
-    final var twist_translation =
-        V_inv.times(VecBuilder.fill(transform.getX(), transform.getY(), transform.getZ()));
-
-    return new Twist3d(
-        twist_translation.get(0, 0),
-        twist_translation.get(1, 0),
-        twist_translation.get(2, 0),
-        rvec.get(0, 0),
-        rvec.get(1, 0),
-        rvec.get(2, 0));
+    return twist3dFromDoubleArray(
+        WPIMathJNI.logPose3d(pose3dToDoubleArray(this), pose3dToDoubleArray(end)));
```

Comment:
We can avoid a bunch of memory allocations here.  Pass the 12 doubles as parameters, and have the JNI side function create the Twist3d return object.  That will avoid the to/from double arrays entirely.